src/wrapper/objects/auto_array.rs

```diff
@@ -17,9 +21,9 @@ use crate::{errors::*, objects::JObject, sys, JNIEnv};
 /// The `get` method must return a valid pointer to the beginning of the JNI array.
 ///
 /// The `release` method must not invalidate the `ptr` if the `mode` is [`sys::JNI_COMMIT`].
-pub unsafe trait TypeArray {
+pub unsafe trait TypeArray: Copy {
     /// getter
-    fn get(env: &mut JNIEnv, obj: &JObject, is_copy: &mut jboolean) -> Result<*mut Self>;
+    fn get(env: &mut JNIEnv, array: jarray, is_copy: &mut jboolean) -> Result<*mut Self>;
```

Comment:
This method now takes and uses a raw pointer (`jarray`), and it's publicly accessible, so it should be made `unsafe`.

Reply:
Right. In general it looks like the intent is that this trait is effectively private but the methods should still be marked as `unsafe`.

I wonder if the trait should also be moved into a private module and marked with `#[doc(hidden)]` to further hide the trait as an implementation detail.

Reply:
People still need to see which types implement the trait.

You might want to use the sealed trait pattern here, like this:

```rust
mod type_array_sealed {
	pub unsafe trait Sealed {
		// Move the `TypeArray` methods into this trait.
	}
	
	impl Sealed for jint {
		// Implement the `TypeArray` methods for `jint` here.
	}
}

pub trait TypeArray: type_array_sealed::Sealed {}

impl TypeArray for jint {}
```

This allows users to see that there is a trait and which types implement it, but hides the actual implementation details.

Reply:
Thanks, this was the pattern I was roughly recalling but this is a nice clear illustration. It seems like this would make sense for this trait.